src/SourceConstraints.hs

```diff
@@ -191,3 +203,34 @@ render dynFlags outputable =
     dynFlags
     (ppr outputable)
     (defaultUserStyle dynFlags)
+
+sortedLocated
+  :: forall a . Outputable a
+  => String
+  -> DynFlags
+  -> [Located a]
+  -> Maybe ErrMsg
+sortedLocated name dynFlags nodes = mkWarning <$> violation
+  where
+    mkWarning :: (String, String, Located a) -> ErrMsg
+    mkWarning (_rendered, expected, node) =
+      mkWarnMsg
+        dynFlags
+        (getLoc node)
+        neverQualify
+        (text $ "Unsorted " <> name <> ", expected: " <> expected)
+
+    violation = find testViolation candidates
+
+    testViolation :: (String, String, Located a) -> Bool
+    testViolation (rendered, expected, _node) = rendered /= expected
+
+    candidates :: [(String, String, Located a)]
+    candidates =
+      let rendered = render dynFlags <$> nodes
+      in
+        zip3 rendered (sort rendered) nodes
+
+zip3 :: [a] -> [b] -> [c] -> [(a, b, c)]
+zip3 (a:as) (b:bs) (c:cs) = (a, b, c) : zip3 as bs cs
+zip3 _      _      _      = []
```

Comment:
What about using `zip3` from `Data.List`? It looks like you import that module in this code already.

Reply:
Hah, I did not think haskell has a build in one.